pkg/alertmanager/alertmanager_test.go

```diff
@@ -0,0 +1,92 @@
+package alertmanager
+
+import (
+	"fmt"
+	"net/url"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/prometheus/alertmanager/config"
+	"github.com/prometheus/alertmanager/types"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/testutil"
+	"github.com/prometheus/common/model"
+	"github.com/stretchr/testify/require"
+
+	"github.com/cortexproject/cortex/pkg/util/test"
+)
+
+func TestDispatcherGroupLimits(t *testing.T) {
+	for name, tc := range map[string]struct {
+		alerts           int
+		groupsLimit      int
+		expectedFailures int
+	}{
+		"no limit":   {alerts: 5, groupsLimit: 0, expectedFailures: 0},
+		"high limit": {alerts: 5, groupsLimit: 10, expectedFailures: 0},
+		"low limit":  {alerts: 5, groupsLimit: 3, expectedFailures: 2},
+	} {
+		t.Run(name, func(t *testing.T) {
+			createAlertmanagerAndSendAlerts(t, tc.alerts, tc.groupsLimit, tc.expectedFailures)
+		})
+	}
+}
+
+func createAlertmanagerAndSendAlerts(t *testing.T, alerts, groupsLimit, expectedFailures int) {
+	user := "test"
+
+	reg := prometheus.NewPedanticRegistry()
+	am, err := New(&Config{
+		UserID:          user,
+		Logger:          log.NewNopLogger(),
+		Limits:          &mockAlertManagerLimits{maxDispatcherAggregationGroups: groupsLimit},
+		TenantDataDir:   t.TempDir(),
+		ExternalURL:     &url.URL{Path: "/am"},
+		ShardingEnabled: false,
+	}, reg)
+	require.NoError(t, err)
+	defer am.StopAndWait()
+
+	cfgRaw := `receivers:
+- name: 'prod'
+
+route:
+  group_by: ['alertname']
+  group_wait: 10ms
+  group_interval: 10ms
+  receiver: 'prod'`
+
+	cfg, err := config.Load(cfgRaw)
+	require.NoError(t, err)
+	require.NoError(t, am.ApplyConfig(user, cfg, cfgRaw))
+
+	now := time.Now()
+
+	for i := 0; i < alerts; i++ {
+		inputAlerts := []*types.Alert{
+			{
+				Alert: model.Alert{
+					Labels:       model.LabelSet{"alertname": model.LabelValue(fmt.Sprintf("Alert-%d", i))},
```

Comment:
(non blocking nit) I was thinking whether we should we test the case of having multiple alerts per group? i.e. This particular limit shouldn't be hit with a much higher number of alerts than groups. Maybe we create 2x alerts per group?

Reply:
Sounds good, I'll do that.